File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/ImageViewPanning.java
package com.productions.gizzmoo.pokemonpuzzleleague;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

/**
 * Created by Chrystian on 6/12/2018.
 */

public class ImageViewPanning extends View {
    private final int backgroundImageResID = R.drawable.menu_background;

    Context mContext;
    private Bitmap mImageBitmap;
    private Rect mScreenDimensions = new Rect();
    private Rect mProportionalRect = new Rect();

    private int mFramesNeedFullView;
    private int mCurrentFrame;
    private boolean mGoingRight;

    public ImageViewPanning(Context context) {
        super(context);
        init(context);
    }

    public ImageViewPanning(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public ImageViewPanning(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(Context context) {
        mContext = context;
        mImageBitmap = BitmapFactory.decodeResource(mContext.getResources(), backgroundImageResID);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        setMeasuredDimension(widthMeasureSpec, heightMeasureSpec);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        mScreenDimensions.set(0, 0, getMeasuredWidth(), getMeasuredHeight());
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        float scalingFactor = (float)mScreenDimensions.bottom / mImageBitmap.getHeight();
        float scaledWidth = scalingFactor * mImageBitmap.getWidth();
        float percentWidth =  mScreenDimensions.right / scaledWidth;
        int widthToShow = (int)(mImageBitmap.getWidth() * percentWidth);

        float frameMoveSize = (float)(mImageBitmap.getWidth() - widthToShow) / mFramesNeedFullView;

        mProportionalRect.set((int)(frameMoveSize * mCurrentFrame), 0, (int)(frameMoveSize * mCurrentFrame) + widthToShow, mImageBitmap.getHeight());
        canvas.drawBitmap(mImageBitmap, mProportionalRect, mScreenDimensions, null);
    }

    public void framesToLookAtView(int frames) {
        mFramesNeedFullView = frames;
    }

    public void setCurrentFrame(int currentFrame) {
        if (currentFrame < 0) {
            currentFrame = 0;
        } else if (currentFrame > mFramesNeedFullView) {
            currentFrame = mFramesNeedFullView;
        }

        mCurrentFrame = currentFrame;
    }

    public void setDirectionRight(boolean isGoingRight) {
        mGoingRight = isGoingRight;
    }

    public void moveFrame() {
        if (mGoingRight && mCurrentFrame >= mFramesNeedFullView) {
            mGoingRight = false;
            mCurrentFrame--;
        } else if (!mGoingRight && mCurrentFrame < 1) {
            mGoingRight = true;
            mCurrentFrame++;
        } else if (mGoingRight) {
            mCurrentFrame++;
        } else {
            mCurrentFrame--;
        }
    }

    public int getCurrentImageFrame() {
        return mCurrentFrame;
    }

    public boolean isPanningRight() {
        return mGoingRight;
    }
}

File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/ImageViewPanning.kt
package com.productions.gizzmoo.pokemonpuzzleleague

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Rect
import android.util.AttributeSet
import android.view.View

class ImageViewPanning(context: Context, attrs: AttributeSet? = null) : View(context, attrs) {
    private val backgroundImageResID = R.drawable.menu_background

    private var imageBitmap: Bitmap = BitmapFactory.decodeResource(context.resources, backgroundImageResID)
    private val screenDimensions = Rect()
    private val proportionalRect = Rect()

    var framesNeededForView: Int = 0
    var currentFrame: Int = 0
        private set
    var direction: Direction = Direction.None

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        setMeasuredDimension(widthMeasureSpec, heightMeasureSpec)
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        screenDimensions.set(0, 0, measuredWidth, measuredHeight)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val scalingFactor = screenDimensions.bottom.toFloat() / imageBitmap.height
        val scaledWidth = scalingFactor * imageBitmap.width
        val percentWidth = screenDimensions.right / scaledWidth
        val widthToShow = (imageBitmap.width * percentWidth).toInt()

        val frameMoveSize = (imageBitmap.width - widthToShow).toFloat() / framesNeededForView

        proportionalRect.set((frameMoveSize * currentFrame).toInt(), 0, (frameMoveSize * currentFrame).toInt() + widthToShow, imageBitmap.height)
        canvas.drawBitmap(imageBitmap, proportionalRect, screenDimensions, null)
    }

    fun setCurrentFrame(frame: Int) {
        currentFrame = when {
            frame < 0 -> 0
            frame > framesNeededForView -> framesNeededForView
            else -> frame
        }
    }


    fun moveFrame() {
        if (direction == Direction.Right && currentFrame >= framesNeededForView) {
            direction = Direction.Left
            currentFrame--
        } else if (direction == Direction.Left && currentFrame < 1) {
            direction = Direction.Right
            currentFrame++
        } else if (direction == Direction.Right) {
            currentFrame++
        } else {
            currentFrame--
        }
    }
}